### PR TITLE
Build meta: fixes and cleanups

### PR DESCRIPTION
**Pull request by xoviat**
Attached are some minor cleanups and fixes to the build_meta module that I added in #1143.

### CODE REVIEW

**Comment by xoviat** [comment]
Okay, this should be ready to go.

File: setuptools/build_meta.py [hunk 1]
```diff
@@ -64,11 +66,13 @@ def patch(cls):
 def _run_setup(setup_script='setup.py'):
     # Note that we can reuse our build directory between calls
     # Correctness comes first, then optimization later
+    _initialize_master_working_set()
```
Comment (benoit-pierre) [inline setuptools/build_meta.py]: This is not necessary anymore, no?

**Comment by jaraco** [comment]
This PR needs more detail. In particular, it needs a bug report or comment describing the undesirable behavior this is correcting. I see some description in the commit message of the first commit (ffb2e69)... and I'd like to know more. In particular, what are the two setuptools versions under which different entry points are loaded?  I feel like I've encountered some of these issues before and want to be sure we're coordinating approaches.

**Comment by benoit-pierre** [comment]
Related: https://github.com/pypa/setuptools/issues/1174.

Adding 1 or 2 tests for the change to `_run_setup` would be great, check if those examples can be run:
```python
variable = True
def function():
  return variable
```
```python
assert __name__ == '__main__'
```

**Comment by xoviat** [comment]
@benoit-pierre Can you explain the local() and __main__ thing here?

**Comment by xoviat** [comment]
>  In particular, it needs a bug report or comment describing the undesirable behavior this is correcting. 

In particular, some setup scripts ([particularly here](https://github.com/pypa/pip/blob/66c53519c2629ec784659fba85c5a84c88144663/tests/data/packages/HackedEggInfo/setup.py)) failed because the `__name__` attribute wasn't set to `__main__`. In addition, passing `locals()` executes the script in the context of the function, which doesn't pass in the classes defined in the module as part of the `setup.py` namespace.

The other change with respect to `dist_info` is a minor cleanup that reduces the number of times that files are copied and, in the case of a read-only directory with a `setup.py`, doesn't fail trying to create `.egg-info`. 

The `pkg_resources` change has been dropped because it actually isn't necessary and I am afraid it would bog down this PR.

**Comment by xoviat** [comment]
Tests have been added. 

@jaraco Hopefully I've done the right thing.

**Comment by xoviat** [comment]
Also many thanks to @benoit-pierre for the excellent PR reviews.

**Comment by xoviat** [comment]
@jaraco ping.

**Comment by jaraco** [comment]
Released as v36.7.3

**Comment by xoviat** [comment]
Thanks. It looks like the pip side has gone off the rails though.

**Pull request merged**